File: ios/JSBridge/CrossJS.m
//
//  CrossJS.m
//  JSBridge
//
//  Created by Wilckerson Ganda on 06/08/14.
//  Copyright (c) 2014 Wilckerson Ganda. All rights reserved.
//

#import "CrossJS.h"

@implementation CrossJS

//Singlethon Pattern
static CrossJS* singleInstance;

+(CrossJS*)Instance
{
    if(singleInstance == nil){
        singleInstance = [[super allocWithZone:NULL] init];
    }
    
    return singleInstance;
}

JSContext* context;

- (id)init {
    if ( (self = [super init]) ) {
        // your custom initialization
        
        //Iniciando o contexto do JavaScriptCore
        context = [[JSContext alloc] init];
        
        //Interceptando os erros de JavaScript e exibindo no console
        [context setExceptionHandler:^(JSContext *context, JSValue *value) {
            //NSLog(@"JS Exception: %@", value);
            [NSException raise: nil format:@"JS Exception: %@",value];

        }];
        
        //Mapeando a funcao console.log para o console Nativo
        [self setJSVariable:@"console.log" nativeValue: ^(NSString *message) {
            NSLog(@"JS LOG: %@", message);
        }];
//        [context evaluateScript:@"var console = {};"]; // make console an empty object
//        context[@"console"][@"log"] = ^(NSString *message) {
//            NSLog(@"%@", message);
//        };
    }
    return self;
}

-(JSContext*) getContext_JavaScriptCore
{
    return context;
}

-(NSString*)loadJSFile: (NSString*) fileName
{
    NSString *jsCode;
    
    //Separando o nome da extensao
    NSRange range = [fileName rangeOfString: @"." options:NSBackwardsSearch];
    NSString* extension = [fileName substringFromIndex: range.location];
    NSString* name = [fileName substringToIndex: range.location];
    
    //Verificando se a extensao é .js
    if(![extension  isEqual: @".js"]){
        [NSException raise: @"Nao é um arquivo JS" format:nil];
    }
    
    //Lendo o arquivo JS
    NSString *filePath = [[NSBundle mainBundle] pathForResource:name ofType:@"js"];
    if (filePath) {
        
        jsCode = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    }
    else
    {
        [NSException raise: @"Erro ao ler arquivo JS" format:@"Nao foi possivel ler o arquivo %@", fileName ];
    }
    
    return jsCode;
}

-(void) loadExecuteJSFile: (NSString*) fileName
{
    NSString* jsCode = [self loadJSFile: fileName];
    [context evaluateScript: jsCode];
}

-(void) setJSVariable: (NSString*)variablePath nativeValue: (id)value
{
    //Verificando se esta chamando variaveis encadeadas. Ex: obj.prop1.prop2
    NSArray* arrayVariables = [variablePath componentsSeparatedByString: @"."];
    
    if(arrayVariables.count <= 1){
        
        //Chamada simples
        context[variablePath] = value;
    }
    else
    {
        //Chamada encadeada
        int i = 1;
        JSValue* currentVariable;
        
        for (NSString* variable in arrayVariables) {
            
            currentVariable = context[variable];
           
            if( [currentVariable isUndefined])
            {
                context[variable] = [[NSObject alloc] init];
                currentVariable = context[variable] ;
            }
            
            //Se chegou na ultima variavel, executa
            if(i == arrayVariables.count-1)
            {
                NSString* finalVariable = [arrayVariables objectAtIndex: i];
                [currentVariable setValue:value forProperty: finalVariable];
                break;
            }
            
            i++;
        }
    }
    
    
}

-(JSValue*) executeJS: (NSString*) jsCode
{
    JSValue* result = [context evaluateScript:jsCode];
    return result;
}

@end;

File: android/CrossJSLib/src/com/crossjs/CrossJS.java
package com.crossjs;

import java.io.IOException;
import java.io.InputStream;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;

public class CrossJS {

	// Singlethon Patter
	static CrossJS singleInstance;

	public static CrossJS getInstance(Context context) {

		if (singleInstance == null) {
			singleInstance = new CrossJS(context);
		}

		return singleInstance;
	}

	WebView webView;
	Context context;

	@SuppressLint("SetJavaScriptEnabled")
	private CrossJS(Context context) {

		this.context = context;

		// Iniciando o WebView
		webView = new WebView(context);

		// Permitindo javascript
		webView.getSettings().setJavaScriptEnabled(true);

		// Mapeando os erros e as requisições de scripts
		webView.setWebViewClient(new WebViewClient() {

			@Override
			public WebResourceResponse shouldInterceptRequest(WebView view, String url) {

				WebResourceResponse result = interceptWebViewRequest(url);

				return (result != null) ? result : super.shouldInterceptRequest(view, url);
			}

			@Override
			public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {

				String msg = String.format("CrossJS Error, Code: %d, Description: %s, FailingURL: %s", errorCode, description, failingUrl);
				Log.e("CrossJS", msg);

				super.onReceivedError(view, errorCode, description, failingUrl);
			}
		});

		// registerWrappers();
	}

	// private void registerWrappers() {
	//
	// // Mapeando a funcao console.log para o console Nativo
	// setJSVariable("console", new LogWrapper());
	//
	// // Mapeando o objeto http para a implementacao nativa
	// setJSVariable("$http", new HttpWrapper(context, scope));
	//
	// }}

	public WebView getContext_WebView() {
		return webView;
	}

	private void registerScriptInWebView(String scriptFilePath) {

		webView.loadUrl("javascript:(function() {" + "var parent = document.getElementsByTagName('head').item(0);" + "var script = document.createElement('script');"
				+ "script.type = 'text/javascript';" + "script.src = 'asset://" + scriptFilePath + "';" + "parent.appendChild(script)" + "})()");

	}

	private WebResourceResponse interceptWebViewRequest(String url) {
		// Como a webView nao consegue ter acesso diretamente para arquivos
		// locais, esse codigo intercepta a requisição para fazer a leitura
		// do arquivo javascript da pasta assets e retorna o conteudo do arquivo
		// como conteudo da requisição HTTP
		String scheme = "asset://";

		if (url.startsWith(scheme)) {
			try {

				String scriptFile = url.replace("asset://", "");

				InputStream input = context.getAssets().open(scriptFile);

				return new WebResourceResponse("text/javascript", "utf-8", input);
			} catch (IOException e) {
				e.printStackTrace();
				return null;
			}
		}
		return null;
	}

	public void loadJSFile(String filePath) throws Exception {

		// Separando o nome da extensao
		String extension = "";

		int i = filePath.lastIndexOf('.');
		if (i > 0) {
			extension = filePath.substring(i);
		} else {
			String msg = "Invalid File.";
			Log.e("loadJSFile", msg);
			throw new Exception(msg);
		}

		// Verificando se a extensão é .js
		if (!extension.equals(".js")) {
			String msg = "This is not a JavaScript file.";
			Log.e("loadJSFile", msg);
			throw new Exception(msg);
		}

		// Registrando na WebView
		registerScriptInWebView(filePath);
	}

	public void setJSVariable(String variablePath, Object nativeValue) {

		// Verificando se esta chamando variaveis encadeadas. Ex:
		// obj.prop1.prop2
		// String[] arrayVariables = variablePath.split("\\.");

		// if (arrayVariables.length <= 1) {

		// Chamada simples
		webView.addJavascriptInterface(nativeValue, variablePath);
		// } else {
		//
		// // Chamada encadeada
		// ScriptableObject currentValue;
		//
		// for (int i = 0; i < arrayVariables.length; i++) {
		//
		// String variable = arrayVariables[i];
		// currentValue = (ScriptableObject) scope.get(variable, null);
		//
		// if (currentValue.isEmpty()) {
		//
		// ScriptableObject.putProperty(currentValue, variable,
		// context.javaToJS(new Object(), scope));
		// currentValue = (ScriptableObject) scope.get(variable, null);
		// }
		//
		// // Se chegou na ultima variavel, executa
		// if (i + 1 == arrayVariables.length - 1) {
		// String finalVariable = arrayVariables[i + 1];
		//
		// currentValue.put(finalVariable, currentValue,
		// context.javaToJS(nativeValue, scope));
		// break;
		// }
		// }
		// }
	}

	public Object executeJS(String jsCode) {
		try {
			webView.loadUrl("javascript:" + jsCode);
			// TODO: Retornar o valor do código executado

		} catch (Exception ex) {

			Log.e("executeJS", ex.getMessage());
		}
		return null;
	}
}
